### PR TITLE
Add pmix variant to openmpi package

### PR DESCRIPTION
**Pull request by MisterFruits**
Allows to compile openmpi against pmix, relates to '--with-pmix=' configure option.

See #19925

### CODE REVIEW

**Comment by hppritcha** [comment]
👍

**Comment by MisterFruits** [comment]
I saw that the gitlabci failed due to some connection refused at the end of the job (it doesn't seem to relate to modifications I made in the PR)

```
ERROR: Job failed (system failure): error dialing backend: dial tcp 192.168.58.124:10250: connect: connection refused
```

Is there an action I can do?

**Comment by hppritcha** [comment]
@alalazo could you rerun the gitlabci job?

File: var/spack/repos/builtin/packages/openmpi/package.py [hunk 1]
```diff
@@ -310,13 +311,16 @@ class Openmpi(AutotoolsPackage):
...
+    depends_on('pmix', when='+pmix')
 
     depends_on('openssh', type='run')
 
     # CUDA support was added in 1.7
     conflicts('+cuda', when='@:1.6')
     # PMI support was added in 1.5.5
     conflicts('+pmi', when='@:1.5.4')
+    # PMIx support was added in 2.0.0
+    conflicts('+pmix', when='@:1.999.999')
```
Comment (haampie) [inline var/spack/repos/builtin/packages/openmpi/package.py]:
```suggestion
    conflicts('+pmix', when='@:1')
```

**Review by haampie — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/openmpi/package.py [hunk 2]
```diff
@@ -641,6 +649,11 @@ def configure_args(self):
                     spec['slurm'].prefix))
             else:
                 config_args.extend(self.with_or_without('pmi'))
+            if spec.satisfies('+pmix'):
+                config_args.append('--with-pmix={0}'.format(
+                    spec['pmix'].prefix))
+            else:
+                config_args.extend(self.with_or_without('pmix'))
```
Comment (haampie) [inline var/spack/repos/builtin/packages/openmpi/package.py]:
```suggestion
            config_args += self.with_or_without('pmix', activation_value='prefix')
```

**Review by haampie — CHANGES_REQUESTED** [review]

**Review by haampie — APPROVED** [review]

**Comment by dylex** [comment]
This breaks the build for openmpi 2, as pmix is required, but building with external pmix also requires external libevent, which isn't enabled for <4.  It would be nice to have an option to use the previous default of using the internal pmix (rather than disabling it completely), or even to make this the default for <4 as it is for libevent.

**Pull request merged**